cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/ssa0/SsaInternals.qll

```diff
@@ -113,58 +53,27 @@ abstract private class DefOrUseImpl extends TDefOrUseImpl {
   /** Gets the location of this element. */
   abstract Cpp::Location getLocation();
 
-  abstract Instruction getBase();
+  abstract BaseSourceVariableInstruction getBase();
 
   final BaseSourceVariable getBaseSourceVariable() {
-    exists(IRVariable var |
-      result.(BaseIRVariable).getIRVariable() = var and
-      instructionHasIRVariable(this.getBase(), var)
-    )
-    or
-    result.(BaseCallVariable).getCallInstruction() = this.getBase()
+    result = this.getBase().getBaseSourceVariable()
   }
 
   /** Gets the variable that is defined or used. */
   final SourceVariable getSourceVariable() {
     exists(BaseSourceVariable v |
-      sourceVariableHasBaseAndIndex(result, v) and
-      defOrUseHasSourceVariable(this, v)
+      result.getBaseVariable() = v and
+      this.getBaseSourceVariable() = v
     )
```

Comment:
Doesn't the following suffice here, given the result type of `getBaseSourceVariable`?
```codeql
result.getBaseVariable() = this.getBaseSourceVariable()
```

Reply:
Oh, good point. Yeah, we can totally do that now. In fact, since `SourceVariable::getBaseVariable()` is actually just the identity we could even do `result = this.getBaseSourceVariable()`. But let's not assume that this will be true forever. So I'll go with your suggestion 👍.